lib/experimental/auto-inserting-blocks.php

```diff
@@ -113,6 +129,36 @@ function gutenberg_register_auto_inserted_blocks( $settings, $metadata ) {
 		$settings['auto_insert'][ $anchor_block_name ] = $mapped_position;
 	}
 
+	// Copied from `get_block_editor_server_block_settings()`.
+	$fields_to_pick = array(
+		'api_version'      => 'apiVersion',
+		'title'            => 'title',
+		'description'      => 'description',
+		'icon'             => 'icon',
+		'attributes'       => 'attributes',
+		'provides_context' => 'providesContext',
+		'uses_context'     => 'usesContext',
+		'selectors'        => 'selectors',
+		'supports'         => 'supports',
+		'category'         => 'category',
+		'styles'           => 'styles',
+		'textdomain'       => 'textdomain',
+		'parent'           => 'parent',
+		'ancestor'         => 'ancestor',
+		'keywords'         => 'keywords',
+		'example'          => 'example',
+		'variations'       => 'variations',
+	);
+	$fields_to_pick['auto_insert'] = 'autoInsert';
+
+	$exposed_settings = array_intersect_key( $settings, $fields_to_pick );
+
+	// TODO: Make work for blocks registered via direct call to gutenberg_register_auto_inserted_block().
+	wp_add_inline_script(
+		'wp-blocks',
+		'wp.blocks.unstable__bootstrapServerSideBlockDefinitions(' . wp_json_encode( array( $inserted_block_name => $exposed_settings ) ) . ');',
```

Comment:
That's interesting that you went that far and you updated this JS call that share block definitions from the server. In the past, we were taking advantage of the fact that `block.json` gets read also by JavaScript, and that allows polyfilling it without recreating PHP logic from core:

https://github.com/WordPress/gutenberg/blob/4a24e8560cb5c277ea90631ad81e3453933bc27e/packages/blocks/src/api/registration.js#L149-L182

We can keep it as is here. I wanted to share the alternative approach used in the past.

Reply:
I tried that at first, but I'm afraid that doesn't work. IIUC, the mechanism you're referencing can only work for block properties that are well-known on the server. It might work fine as a client-side shim for a property that has been stabilized and added on the server, but it doesn't seem to work for block properties that haven't 😕 

The problem is that the server injects one [`wp.blocks.unstable__bootstrapServerSideBlockDefinitions` call](https://github.com/WordPress/wordpress-develop/blob/d4b496acab95f9ddd5a87c8224ca001c2131de1a/src/wp-admin/site-editor.php#L128-L132) which includes only the attributes [from a given (and unfilterable) allowlist](https://github.com/WordPress/wordpress-develop/blob/12a0205829c1411f16b4e216a713a153bcc62e2f/src/wp-admin/includes/post.php#L2216-L2234), and subsequent "actual" client-side calls [will never override any existing block definitions](https://github.com/WordPress/gutenberg/blob/4a24e8560cb5c277ea90631ad81e3453933bc27e/packages/blocks/src/api/registration.js#L146-L147). So the workaround I found was to inject a `wp.blocks.unstable__bootstrapServerSideBlockDefinitions` call for blocks with an `__experimentalAutoInserted` property early in the server, so that the existing "other" `wp.blocks.unstable__bootstrapServerSideBlockDefinitions` injected call wouldn't override those.

It's kind of a nasty hack, but it's the only way I found to solve a fairly tricky problem that looked like it might entirely block this PR altogether (and would require a change in Core) for a while 😅 (For more context, see [here](https://github.com/WordPress/gutenberg/pull/52969#issuecomment-1653214283) and [here](https://github.com/WordPress/gutenberg/pull/52969#issuecomment-1653255323).) Of course, we'll want to tackle this quite differently when merging the code into Core.

Reply:
It would work for properties defined in `block.json`, but you would have to whitelist it also in another place:

https://github.com/WordPress/gutenberg/blob/3894324874c14ca2f585a06e76a12036ae2c0c54/packages/blocks/src/api/registration.js#L205-L222

It's not included as of today.

Reply:
I thought I'd tried adding it there as well, but I guess I must've missed that one!

Reply:
I quickly tried the following:

```diff
diff --git a/packages/blocks/src/api/registration.js b/packages/blocks/src/api/registration.js
index bf866b7a21..702582ff64 100644
--- a/packages/blocks/src/api/registration.js
+++ b/packages/blocks/src/api/registration.js
@@ -180,6 +180,17 @@ export function unstable__bootstrapServerSideBlockDefinitions( definitions ) {
                                serverSideBlockDefinitions[ blockName ].selectors =
                                        definitions[ blockName ].selectors;
                        }
+
+                       if (
+                               serverSideBlockDefinitions[ blockName ]
+                                       .__experimentalAutoInsert === undefined &&
+                               definitions[ blockName ].__experimentalAutoInsert
+                       ) {
+                               serverSideBlockDefinitions[
+                                       blockName
+                               ].__experimentalAutoInsert =
+                                       definitions[ blockName ].__experimentalAutoInsert;
+                       }
                        continue;
                }
 
@@ -219,6 +230,7 @@ function getBlockSettingsFromMetadata( { textdomain, ...metadata } ) {
                'styles',
                'example',
                'variations',
+               '__experimentalAutoInsert',
        ];
 
        const settings = Object.fromEntries(
```

but that didn't work 😬

I guess something else is still missing; probably the property needs to be allowlisted elsewhere. (Wouldn't it need to be in the schema after all if we need to parse it from `block.json`?)

Anyway, maybe we can address this separately -- I'd rather not make it a blocker for this PR 🙂